internal/controllers/topology/cluster/desired_state.go

```diff
@@ -873,6 +890,284 @@ func isMachineDeploymentDeferred(clusterTopology *clusterv1.Topology, mdTopology
 	return false
 }
 
+// computeMachinePools computes the desired state of the list of MachinePools.
+func (r *Reconciler) computeMachinePools(ctx context.Context, s *scope.Scope) (scope.MachinePoolsStateMap, error) {
+	machinePoolsStateMap := make(scope.MachinePoolsStateMap)
+	for _, mpTopology := range s.Blueprint.Topology.Workers.MachinePools {
+		desiredMachinePool, err := computeMachinePool(ctx, s, mpTopology)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to compute MachinePool for topology %q", mpTopology.Name)
+		}
+		machinePoolsStateMap[mpTopology.Name] = desiredMachinePool
+	}
+	return machinePoolsStateMap, nil
+}
+
+// computeMachinePool computes the desired state for a MachinePoolTopology.
+// The generated machinePool object is calculated using the values from the machinePoolTopology and
+// the machinePool class.
+func computeMachinePool(_ context.Context, s *scope.Scope, machinePoolTopology clusterv1.MachinePoolTopology) (*scope.MachinePoolState, error) {
+	desiredMachinePool := &scope.MachinePoolState{}
+
+	// Gets the blueprint for the MachinePool class.
+	className := machinePoolTopology.Class
+	machinePoolBlueprint, ok := s.Blueprint.MachinePools[className]
+	if !ok {
+		return nil, errors.Errorf("MachinePool class %s not found in %s", className, tlog.KObj{Obj: s.Blueprint.ClusterClass})
+	}
+
+	var machinePoolClass *clusterv1.MachinePoolClass
+	for _, mpClass := range s.Blueprint.ClusterClass.Spec.Workers.MachinePools {
+		mpClass := mpClass
+		if mpClass.Class == className {
+			machinePoolClass = &mpClass
+			break
+		}
+	}
+	if machinePoolClass == nil {
+		return nil, errors.Errorf("MachinePool class %s not found in %s", className, tlog.KObj{Obj: s.Blueprint.ClusterClass})
+	}
+
+	// Compute the bootstrap config.
+	currentMachinePool := s.Current.MachinePools[machinePoolTopology.Name]
+	var currentBootstrapConfigRef *corev1.ObjectReference
+	if currentMachinePool != nil && currentMachinePool.BootstrapObject != nil {
+		currentBootstrapConfigRef = currentMachinePool.Object.Spec.Template.Spec.Bootstrap.ConfigRef
+	}
+	var err error
+	desiredMachinePool.BootstrapObject, err = templateToObject(templateToInput{
+		template:              machinePoolBlueprint.BootstrapTemplate,
+		templateClonedFromRef: contract.ObjToRef(machinePoolBlueprint.BootstrapTemplate),
+		cluster:               s.Current.Cluster,
+		namePrefix:            bootstrapTemplateNamePrefix(s.Current.Cluster.Name, machinePoolTopology.Name),
+		currentObjectRef:      currentBootstrapConfigRef,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to compute bootstrap object for topology %q", machinePoolTopology.Name)
+	}
+
+	bootstrapTemplateLabels := desiredMachinePool.BootstrapObject.GetLabels()
```

Comment:
To stay consistent in naming: we should rename this to `bootstrapObjectLabels`?


```suggestion
	bootstrapObjectLabels := desiredMachinePool.BootstrapObject.GetLabels()
```